src/Data/Singletons/Promote/Defun.hs

```diff
@@ -340,279 +386,496 @@ defunctionalize name m_fixity m_arg_tvbs' m_res_kind' = do
+this:
+
+  type (*>@#@$$$) :: f a -> f b -> f b
+  type family (*>@#@$$$) x y where
+    (*>@#@$$$) x y = (*>) x y
+
+Like before, the full kind of (*>@#@$$$) is generalized to be
+`forall {k} (f :: k -> Type) (a :: k) (b :: k)`. The difference is that the
+type family equation *matches* on `k` such that the equation will only trigger
+if `k` is equal to `Type`. Note that (*>@#@$) and (*>@#@$$) employ a similar
+trick, as the GADT constructors for each of those symbols *also* constrain `k`
+to be `Type`.
+
+Alas, the type family approach is strictly less powerful than the type synonym
+approach. Consider the following code:
+
+  $(singletons [d|
+    data Nat = Z | S Nat
+
+    natMinus :: Nat -> Nat -> Nat
+    natMinus Z     _     = Z
+    natMinus (S a) (S b) = natMinus a b
+    natMinus a     Z     = a
+    |])
+
+Among other things, this will generate the following declarations:
+
+  type ZSym0 :: Nat
+
+  type NatMinus :: Nat -> Nat -> Nat
+  type family NatMinus x y where
+    NatMinus Z     _     = ZSym0
+    NatMinus (S a) (S b) = NatMinus a b
+    NatMinus a     Z     = a
+
+  sNatMinus :: SNat x -> SNat y -> SNat (NatMinus x y)
+  sNatMinus SZ      _       = SZ
+  sNatMinus (SS sA) (SS sB) = sNatMinus sA sB
+  sNatMinus sA      SZ      = sA
+
+Shockingly, this will either succeed or fail to compile depending on whether
+ZSym0 is a type synonym or a type family. If ZSym0 is a type synonym, then
+the first and third equations of NatMinus will be compatible (since GHC will
+be able to infer that Z ~ ZSym0), which is what allows the third equation of
+sNatMinus to typecheck. If ZSym0 is a type family, however, then the third
+equation of NatMinus will be incompatible with the first, which will cause
+the third equation of sNatMinus to fail to typecheck:
+
+  error:
+      • Could not deduce: NatMinus x 'Z ~ x
+        from the context: y ~ 'Z
+          bound by a pattern with constructor: SZ :: SNat 'Z,
+                   in an equation for ‘sNatMinus’
+        ‘x’ is a rigid type variable bound by
+          the type signature for:
+            sNatMinus :: forall (x :: Nat) (y :: Nat).
+                         SNat x -> SNat y -> SNat (NatMinus x y)
+        Expected type: SNat (NatMinus x y)
+          Actual type: SNat x
+      • In the expression: sA
+        In an equation for ‘sNatMinus’: sNatMinus sA SZ = sA
+      • Relevant bindings include
+          sA :: SNat x
+          sNatMinus :: SNat x -> SNat y -> SNat (NatMinus x y)
+
+One could work around the issue by tweaking the third equation of natMinus
+slightly:
+
+  $(singletons [d|
+    ...
+
+    natMinus :: Nat -> Nat -> Nat
+    natMinus Z       _     = Z
+    natMinus (S a)   (S b) = natMinus a b
+    natMinus a@(S _) Z     = a
+    |])
+
+But I would generally prefer to avoid having the user add extraneous pattern
+matches when possible. Given the choice between expressiveness and SAKs, I give
+the edge to expressiveness.
```

Comment:
But we explicitly don't support overlapping equations, and the killer example uses those. I don't necessarily think it's worth fixing this now, but if the lack of a SAK here causes problems, I think this is viable.

Reply:
The lack of a SAK (hey, that rhymes!) certainly doesn't cause problems now, but it does create a major hurdle if we ever want to fully support promoting visible type applications. This is because `f` is always promoted to `FSym0`, even if it isn't applied to any arguments. In order to promote `f @x` to `FSym0 @x`, it is imperative that the order of kind variables in `FSym0` is the same as in `f`.

Moreover, this isn't even the only hurdle we'd have to clear in order to promote visible type application. See also `Note [Promoted class methods and kind variable ordering]` and `Note [No SAKs for let-decs with local variables]` in this patch. These problems discussed in those Notes don't have a clear solution, which leads me to believe that fixing chunk (3) in https://github.com/goldfirere/singletons/issues/378#issuecomment-528519679 is a pipe dream (at least for now).

In light of all this, it seems silly to change saturated defunctionalization symbols to type families (which induces a breaking change) when it won't even get us that much closer to the stated goal of promoting visible type applications. Let me know if you feel strongly otherwise, but I'm disinclined to rock the boat without first charting a clear path to shore.

Reply:
What's worse than a mixed metaphor? A pseudo-mixed one, which you successfully pulled off in your last sentence. I'll have to remember that one.

We're in agreement here. My comment was merely to say that we should not unduly prioritize accepting this example.